mlir/lib/Dialect/MIOpen/Transforms/AlignTiling.cpp

```diff
@@ -95,81 +96,88 @@ template <typename T> struct MILARewritePattern : public OpRewritePattern<T> {
+    auto twinp = miTWCopy.getOperand(0);
+    auto miTransform = twinp.getDefiningOp<miopen::TransformOp>();
+    auto miTransformInp = miTransform->getOperand(0);
+    auto miAlloc = miTransformInp.getDefiningOp<miopen::GpuAllocOp>();
+
+    // 1. clone reg alloc + transform
+    BlockAndValueMapping cloningMap;
+    auto nAlloc = b.clone(*miAlloc, cloningMap);
+    cloningMap.map(miTransform->getOperand(0), nAlloc->getResult(0));
+    auto nTransform = b.clone(*miTransform, cloningMap);
+
+    // 2. clone twcopy for <addend> -> regs
+    cloningMap.map(miTWCopy->getOperand(0), inp);
+    cloningMap.map(miTWCopy->getOperand(1), nTransform->getResult(0));
+    auto nTWCopy = b.clone(*miTWCopy, cloningMap);
+
+    // 3. swap input coords with output coords
+    for (uint i = 0; i < 5; ++i) {
+      uint inIdx = 2 + i;
+      uint outIdx = 2 + 5 + i;
+      auto inCoord = nTWCopy->getOperand(inIdx);
+      auto outCoord = nTWCopy->getOperand(outIdx);
+      nTWCopy->setOperand(inIdx, outCoord);
+      nTWCopy->setOperand(outIdx, inCoord);
     }
-    AffineMap outputAffineMap =
-        AffineMap::get(outputDims, 0, ArrayRef<AffineExpr>{expr}, ctx);
-    auto transformedOutputType =
-        MemRefType::get(outputShape, outputElementType, {outputAffineMap});
-
-    llvm::SmallVector<NamedAttribute, 3> transformedNewOutputAttrs;
-    auto transform = b.create<miopen::TransformOp>(
-        loc, transformedOutputType, inp, transformedNewOutputAttrs, true);
-
-    // 2. reduce scope of inp to tile size
-    //      - use subview:
-    //      https://mlir.llvm.org/docs/Dialects/MemRef/#memrefsubview-mlirmemrefsubviewop
-    auto regs = twcopy.getOperand(0);
-    auto regShape = regs.getType().template cast<MemRefType>().getShape();
-    auto regDims = regShape.size();
-    for (uint32_t i = 0; i < regDims; ++i) {
-      Value idx = b.create<IndexCastOp>(loc, twcopy.getOperand(2 + regDims + i),
-                                        b.getIndexType());
-      offsets[i] = idx;
-      sizes[i] = b.getIndexAttr(regShape[i]);
+
+    return nAlloc->getResult(0);
+  }
+
+  Value applyTransforms(PatternRewriter &b, miopen::ThreadwiseCopyOp &miTWCopy,
+                        Value inp, SmallVector<Value, 5> &transforms) const {
+    Value ret = inp;
+    BlockAndValueMapping cloningMap;
+    // 1. clone the same transforms applied to the output memory and
+    //    apply to all other inputs to the linalg.generic
+    for (auto transform : transforms) {
+      assert(transform.hasOneUse());
+      Operation *tcopy;
+      if (auto miTransform = transform.getDefiningOp<miopen::TransformOp>()) {
+        cloningMap.map(miTransform->getOperand(0), ret);
+        tcopy = b.clone(*miTransform, cloningMap);
+      } else if (auto laReshape =
+                     transform.getDefiningOp<memref::ExpandShapeOp>()) {
+        cloningMap.map(laReshape->getOperand(0), ret);
+        tcopy = b.clone(*laReshape, cloningMap);
+      } else {
+        assert(0);
```

Comment:
`miTransform` and `laReshape` logic here seem duplicated. Can we use `isa<>` or `dyn_cast<>` and make the two branches into 1?

and, `assert(0)` fails the process silently. We want to make some noises here.